torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -301,16 +307,26 @@ c10::intrusive_ptr<OwnerRRef> RRefContext::createOwnerRRef(
       getWorkerId(), genGloballyUniqueId(), type);
 }
 
-c10::intrusive_ptr<OwnerRRef> RRefContext::getOwnerRRef(const RRefId& rrefId) {
+std::shared_ptr<torch::utils::Future<c10::intrusive_ptr<OwnerRRef>>>
+RRefContext::getOwnerRRef(const RRefId& rrefId) {
   std::unique_lock<std::mutex> lock(mutex_);
   const auto iter = owners_.find(rrefId);
   if (iter == owners_.end()) {
     // Scenario (1) RRef is used before it is created
-    ownerCV_.wait(lock, [&] { return owners_.find(rrefId) != owners_.end(); });
-    return c10::static_intrusive_pointer_cast<OwnerRRef>(owners_[rrefId]);
+    const auto pendingOwnerIter = pendingOwners_.find(rrefId);
+    if (pendingOwnerIter == pendingOwners_.end()) {
+      auto futureOwner = std::make_shared<
+          torch::utils::Future<c10::intrusive_ptr<OwnerRRef>>>();
+      pendingOwners_[rrefId] = futureOwner;
+      return futureOwner;
+    } else {
+      return pendingOwnerIter->second;
+    }
   } else {
     // Scenario (2) retrieving an existing RRef
-    return c10::static_intrusive_pointer_cast<OwnerRRef>(iter->second);
+    return std::make_shared<
```

Comment:
I guess this implicitly assumes that the future API marks a future as completed if it's created with a value `T`, although it doesn't seem like this holds true for the ivalue generic future. Not sure if this assumption might introduce bugs when we attempt to unify the futures.